File: frontend/src/services/partService.ts
import api from './api';
import { Part, PageResponse } from '../types';
import { mockStore } from './mockData';

export const partService = {
  getAll: async (search?: string, page = 0, size = 20): Promise<PageResponse<Part>> => {
    try {
      const params = new URLSearchParams();
      if (search) params.append('search', search);
      params.append('page', page.toString());
      params.append('size', size.toString());
      const response = await api.get(`/parts?${params.toString()}`);
      if (response.data && Array.isArray(response.data.content)) {
        return response.data;
      }
      return mockStore.getParts(search, page, size);
    } catch {
      return mockStore.getParts(search, page, size);
    }
  },
  getById: async (id: number): Promise<Part> => {
    try {
      const response = await api.get(`/parts/${id}`);
      if (response.data && response.data.id) return response.data;
      const p = mockStore.getPartById(id);
      if (p) return p;
      throw new Error('Part not found');
    } catch {
      const p = mockStore.getPartById(id);
      if (!p) throw new Error('Part not found');
      return p;
    }
  },
  create: async (data: Partial<Part>): Promise<Part> => {
    try {
      const response = await api.post('/parts', data);
      if (response.data && response.data.id) return response.data;
      return mockStore.createPart(data);
    } catch {
      return mockStore.createPart(data);
    }
  },
  update: async (id: number, data: Partial<Part>): Promise<Part> => {
    try {
      const response = await api.put(`/parts/${id}`, data);
      if (response.data && response.data.id) return response.data;
      return mockStore.updatePart(id, data);
    } catch {
      return mockStore.updatePart(id, data);
    }
  },
  getLowStock: async (): Promise<Part[]> => {
    try {
      const response = await api.get('/parts/low-stock');
      if (Array.isArray(response.data)) return response.data;
      return mockStore.getLowStockParts();
    } catch {
      return mockStore.getLowStockParts();
    }
  },
};

File: backend/src/main/java/com/vertexa/keystone/service/PartService.java
package com.vertexa.keystone.service;

import com.vertexa.keystone.domain.Part;
import com.vertexa.keystone.domain.enums.UserRole;
import com.vertexa.keystone.dto.part.PartRequest;
import com.vertexa.keystone.dto.part.PartResponse;
import com.vertexa.keystone.exception.DuplicateResourceException;
import com.vertexa.keystone.exception.ForbiddenOperationException;
import com.vertexa.keystone.exception.ResourceNotFoundException;
import com.vertexa.keystone.mapper.PartMapper;
import com.vertexa.keystone.repository.PartRepository;
import com.vertexa.keystone.util.SecurityUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class PartService {

    private final PartRepository partRepository;
    private final PartMapper partMapper;
    private final SecurityUtils securityUtils;

    @Value("${keystone.parts.low-stock-threshold:5}")
    private int lowStockThreshold;

    @Transactional(readOnly = true)
    public Page<PartResponse> getAllParts(String search, Pageable pageable) {
        Page<Part> parts;
        if (search != null && !search.isBlank()) {
            parts = partRepository.searchActiveParts(search, pageable);
        } else {
            parts = partRepository.findByActiveTrue(pageable);
        }
        return parts.map(partMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public PartResponse getPartById(Long id) {
        Part part = partRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Part", "id", id));
        return partMapper.toResponse(part);
    }

    @Transactional
    public PartResponse createPart(PartRequest request) {
        UserRole role = securityUtils.getCurrentUserRole();
        if (role != UserRole.MANAGER) {
            throw new ForbiddenOperationException("Only managers can create parts");
        }

        if (partRepository.existsByPartCode(request.getPartCode())) {
            throw new DuplicateResourceException("Part with code '" + request.getPartCode() + "' already exists");
        }

        Part part = Part.builder()
                .partCode(request.getPartCode())
                .name(request.getName())
                .description(request.getDescription())
                .unitCost(request.getUnitCost())
                .availableStock(request.getAvailableStock())
                .active(true)
                .build();

        part = partRepository.save(part);
        log.info("Part created: {} ({})", part.getName(), part.getPartCode());
        return partMapper.toResponse(part);
    }

    @Transactional
    public PartResponse updatePart(Long id, PartRequest request) {
        UserRole role = securityUtils.getCurrentUserRole();
        if (role != UserRole.MANAGER) {
            throw new ForbiddenOperationException("Only managers can update parts");
        }

        Part part = partRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Part", "id", id));

        part.setPartCode(request.getPartCode());
        part.setName(request.getName());
        part.setDescription(request.getDescription());
        part.setUnitCost(request.getUnitCost());
        part.setAvailableStock(request.getAvailableStock());

        part = partRepository.save(part);
        log.info("Part updated: {} ({})", part.getName(), part.getPartCode());
        return partMapper.toResponse(part);
    }

    @Transactional(readOnly = true)
    public List<PartResponse> getLowStockParts() {
        List<Part> lowStockParts = partRepository.findByAvailableStockLessThanEqual(lowStockThreshold);
        return lowStockParts.stream()
                .map(partMapper::toResponse)
                .collect(Collectors.toList());
    }
}
